bsb/morphologies/__init__.py

```diff
@@ -788,6 +821,113 @@ def points(self):
+
+    @property
+    def end(self):
+        """
+        Return the spatial coordinates of the terminal point of this branch.
+        """
+        try:
+            return self._points[-1]
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no ending point") from None
+
+    @property
+    def vector(self):
+        """
+        Return the vector of the axis connecting the start and terminal points.
+        """
+        try:
+            return self.end - self.start
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no vector") from None
+
+    @property
+    def versor(self):
+        """
+        Return the normalized vector of the axis connecting the start and terminal points.
+        """
+        try:
+            return (self.end - self.start) / np.linalg.norm(self.end - self.start)
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no versor") from None
+
+    @property
+    def euclidean_dist(self):
+        """
+        Return the Euclidean distance from the start to the terminal point of this branch.
+        """
+        try:
+            return np.sqrt(np.sum((self.end - self.start) ** 2))
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no Euclidean distance") from None
+
+    @property
+    def path_dist(self):
+        """
+        Return the path distance from the start to the terminal point of this branch,
+        computed as the sum of Euclidean segments between consecutive branch points.
+        """
+        try:
+            return np.sum(np.sqrt(np.sum(self.point_vectors**2, axis=1)))
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no path distance") from None
+
+    @property
+    def max_displacement(self):
+        """
+        Return the max displacement of the branch points from its axis vector.
+        """
+        try:
+            displacements = np.linalg.norm(
+                np.cross(self.versor, (self.points - self.start)), axis=1
+            )
+            return np.max(displacements)
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no displaced points") from None
+
+    @property
+    def fractal_dim(self):
+        """
+        Return the fractal dimension of this branch, computed as the coefficient
+        of the line fitting the log-log plot of path vs euclidean distances of its points.
+        """
+        try:
+            euclidean = np.sqrt(np.sum((self.points - self.start) ** 2, axis=1))
+            path = np.cumsum(np.sqrt(np.sum(self.point_vectors**2, axis=1)))
+            log_e = np.log(euclidean[1:])
+            log_p = np.log(path)
+            if len(self.points) <= 2:
+                return 1.0
+            return np.corrcoef(log_e, log_p)[0][1] * (np.std(log_p) / np.std(log_e))
+        except IndexError:
+            raise EmptyBranchError("Empty branch has no fractal dimension") from None
```

Comment:
Try it like this, where you use the previous `EmptyBranchError` to trigger a new one, and use the properties:

```suggestion
        try:
            log_e = np.log(self.euclidean_dist[1:])
        except EmptyBranchError:
            raise EmptyBranchError("Empty branch has no fractal dimension") from None
        else:
            log_p = np.log(self.path_length)
            return np.corrcoef(log_e, log_p)[0][1] * (np.std(log_p) / np.std(log_e))
```